Archives/VolFile.cpp

```diff
@@ -389,103 +382,102 @@ namespace Archives
-		// -------------------------------------------
+		m_HeaderLength = ReadTag(std::array<char, 4> { 'V', 'O', 'L', ' ' });
 
-		// Check opening header tag ("VOL " tag)
-		// -------------------------------------
-		if ((m_HeaderLength = ReadTag(0, "VOL ")) == -1) {
-			return false;
+		uint32_t volhSize = ReadTag(std::array<char, 4> { 'v', 'o', 'l', 'h' });
+		if (volhSize != 0) {
+			return false; //the size of this section must be 0
 		}
 
-		// Make sure the file is large enough to contain the header
-		if (m_MappedFileSize < m_HeaderLength + 8) {
-			return false;
-		}
+		m_StringTableLength = ReadTag(std::array<char, 4> { 'v', 'o', 'l', 's' });
 
-		// Check next header tag ("volh" tag)
-		// ----------------------------------
-		// Note: the size of this section must be 0
-		if (ReadTag(8, "volh") != 0) {
-			return false;
-		}
-
-		// Check for fileName string table tag ("vols" tag)
-		// ------------------------------------------------
-		if ((m_StringTableLength = ReadTag(16, "vols")) == -1) {
-			return false;
-		}
-
-		m_ActualStringTableLength = *((int*)m_BaseOfFile + 6);	// Store actual length
-		m_StringTable = (char*)m_BaseOfFile + 28;				// Store pointer to table
+		ReadStringTable();
 
 		// Make sure the string table fits in the header (and next header fits too)
 		if (m_HeaderLength < m_StringTableLength + 20) {
 			return false;
 		}
 
-		temp = m_StringTableLength + 24;			// Calculate offset to next section
-		// Check for volume index tag ("voli" tag)
-		// ---------------------------------------
-		// Store the offset to the Index table while it's still calculated
-		m_Index = (IndexEntry*)((char*)m_BaseOfFile + temp + 8);
-		if ((m_IndexTableLength = ReadTag(temp, "voli")) == -1) {
-			return false;
+		m_IndexTableLength = ReadTag(std::array<char, 4> { 'v', 'o', 'l', 'i' });
+		if (m_IndexTableLength > 0) {
+			m_IndexEntries.resize(m_IndexTableLength / sizeof(IndexEntry));
+			archiveFileReader->Read(m_IndexEntries.data(), m_IndexTableLength);
 		}
 
 		// Make sure the index table fits in the header
 		if (m_HeaderLength < m_StringTableLength + m_IndexTableLength + 24) {
 			return false;
 		}
 
-		// Determine the number of index entries
-		// -------------------------------------
 		m_NumberOfIndexEntries = m_IndexTableLength / 14;
+		ReadPackedFileCount();
+
+		return true;
+	}
+
+	void VolFile::ReadStringTable()
+	{
+		uint32_t actualStringTableLength;
+		archiveFileReader->Read(&actualStringTableLength, sizeof(actualStringTableLength));
+
+		std::string charBuffer;
+		charBuffer.resize(actualStringTableLength);
+		archiveFileReader->Read(&charBuffer[0], actualStringTableLength);
+
+		m_StringTable.push_back("");
+		for (std::size_t i = 0; i < charBuffer.size(); i++)
+		{
+			if (charBuffer[i] == '\0') {
+				m_StringTable.push_back("");
+				continue;
+			}
+
+			m_StringTable[m_StringTable.size() - 1].push_back(charBuffer[i]);
+		}
+
+		m_StringTable.erase(m_StringTable.begin() + m_StringTable.size() - 1);
+	}
```

Comment:
I'm rather confused by this function. In particular, if a string starts with `'\0'`, it is already `""`, so I don't think there is a need for special casing.

I'm uncertain why `erase` is being used at the end.

Reply:
The code is adding an empty string whenever a null terminator is encounter. Then it fills the string with characters until the next null terminator is found. 

The final null terminator will cause an empty string to be added to the vector, which must be deleted when the for loop is completed.

There is probably a better way to accomplish this in terms of efficiency and readbility, although the provided code doesn't seem to have any bugs that I've seen.

Reply:
Oh, I get it now. That is a bit obscure. And I had misread the `+` as a `,` in the `erase` call, making it look like it was erasing the whole table just after it had been constructed.

Maybe we can flag this for review at a later point once we've got other higher priority items out of the way.

I think one way to make the code easier to read, is using the null terminated string constructor for `std::string`, rather than scanning and appending characters one at a time. The address of the next string can then be calculated using the length of the previous string plus the null terminator. Then you don't need to append empty placeholder string objects.